### DISCUSSION TITLE
Low performance on Sun SparcStation 5 (due to fragmented file)

### DISCUSSION THREAD

**Discussion by wvalcke** [ZuluSCSI/ZuluSCSI-firmware#312] [Q&A] [answered]
The zuluscsi (2040) card works but has a horribe performance when the system boots from the SD card.
It is for sure at least 10x slower than the HD i used before. 
I tried several things 
MaxSyncSpeed = 10 # Set to 5 or 10 to enable synchronous SCSI mode, 0 to disable
I've set it to 10, 5 and 0, but nothing helps, the performance remains very slow.
Does anyone has some experience with sparcstation and zulusci 2040?

**Comment by aperezbios** [comment]
We're going to need a lot more information about your specific set-up, particularly which OS and version you're using (It could be Solaris or NetBSD). Please provide a copy of your zululog, and specifics about the SD card you're using. Did you source it on your own, or purchase it bundled with a ZuluSCSI?

I've also converted this issue to a GitHub Discussion for now, since there's zero hard data. If we decide it's actionable as an issue that needs resolution on the firmware side, I'll convert it back to an issue.

**Comment by wvalcke** [comment]
Hi, thank you for the reply. 
The information i already have is that currently i tried first with Solaris 8, there the performance was very slow.
I tried already to install Solaris 9, to me it looks a bit better, but still definitely slow, much slower than the HD.
I will turn on debugging on the ZuluScsi, and give detailed info about the Sd card. Log file will follow.
I'll keep you infirmed.
I will also try to install NetBsd (9.3) to see if there is a difference in performance.

**Comment by wvalcke** [comment]
Here some more information : the booted system is Solaris 9, booted into single user mode. 
The used SD card is a SanDisk Extreme 32GB.
In attachment, the comple logfile. I hope something can be seen. 
What i saw is for example : DBG SD card readSector(4232) slow transfer, buffer0x2001F2D0 vs. 0x200269A4
Does this indicate a problem ?
[ZuluLogData.zip](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/12849300/ZuluLogData.zip)

**Comment by wvalcke** [comment]
I did some testing with Solaris 9, when fully booted (which is considerable slower than the HD) i tested the following
dd if=somebigfile of=/dev/null
somebigfile is 88MB
This takes 27 secs -> so roughly 3.2MB/sec.

**Comment by wvalcke** [comment]
More test results : Same sparcstation 5, now booted Solaris 9 from HD. Doing a write test to HD gives 8MB/sec.
The ZuluScsi is 4 times slower than the HD. A write test i did gave about 2MB/sec with the ZuluSCSI.
I expected the ZuluSCSI to be at least performing with the same speed.
When i've some time, i'll try NetBSD 9.3 on the ZuluSCSI.

**Comment by aperezbios** [comment]
@wvalcke would you mind stating what performance you expect? Your ad-hoc testing with dd is going to yield artificially-low numbers unless you specify a reasonably-sized block size (bs=512 being the default). You should test with 4K/8K/32K/64. 

What is the make and model of the baseline drive you're testing against? 2MB/sec at a block size of 512 bytes is about what I'd expect. Unless you share specifics of your methodology, the numbers mean nothing to anyone but you.

**Comment by wvalcke** [comment]
Hi, thank you for the reply.
Sorry for the missing detail. I used a block size of 8192 bytes
Write test was : 
dd if=/dev/zero of=outputfile bs=8192 count=10240 && sync
I will do some testing with block sizes of 32K and 64K as well and let you know the results.
The performance i hope to expect is to have read speeds of about 6-8MB/Sec, and writes up to 4-5 MB/Sec
At this moment the bootup of Solaris 9 and logging into the system is a lot slower when using the ZuluSCSI, compared to the HD.
At first i tried to install Solaris 8 on the ZuluSCSI, install went fine, (took 4-5 hours), but then starting the system is extremely slow. 
It takes you minutes (4-5) before you have a login prompt. 
To have some idea what could be the reason, i tried also Solaris 9, when it installed i had the feeling it was already a bit quicker. But still
it is definitely 2-3 times slower than with the HD. But at least it is possible to login, and start CDE etc,... With Solaris 8 i never went that far, it takes ages to login...

**Reply by aperezbios** [reply]
Please share the make and model of hard drive, as requested previously

**Reply by aperezbios** [reply]
@wvalcke we also need a copy of your zululog.

**Reply by wvalcke** [reply]
The HD is a FUJITSU MAB3091S

**Reply by wvalcke** [reply]
[10ms] Platform: ZuluSCSI RP2040
[10ms] FW Version: 23.09.06-release Sep  7 2023 00:03:27
[11ms] DIP switch settings: debug log 0, termination 1
[11ms] SCSI termination is enabled
[12ms] Flash chip size: 4096 kB
[14ms] SCSI target/disk mode selected by DIP switch, acting as a SCSI disk
[169ms] SD card detected, FAT64 volume size: 30430 MB
[170ms] SD MID: 0x03, OID: 0x53 0x44
[170ms] SD Name: SM32G
[170ms] SD Date: 8/2020
[170ms] SD Serial: 0x1599B6B3
[173ms] Reading configuration from zuluscsi.ini
[174ms] Active configuration:
[175ms] -- SelectionDelay = 255
[175ms] -- EnableUnitAttention = No
[176ms] -- EnableSCSI2 = Yes
[177ms] -- EnableSelLatch = No
[178ms] -- MapLunsToIDs = No
[178ms] -- EnableParity = Yes
[301ms] Finding HDD images in directory /:
[302ms] -- Opening /HD1.img for id:1 lun:0
[303ms] ---- WARNING: file /HD1.img is not contiguous. This will increase read latency.
[303ms] ---- Read prefetch enabled: 8192 bytes
[309ms] -- Platform supports ROM drive up to 3740 kB
[309ms] ---- ROM drive image not detected
[310ms] SCSI ID:1 BlockSize:512 Type:0 Quirks:0 ImageSize:15728640kB
[410ms] Initialization complete!
[256296ms] SCSI ID 1 negotiated synchronous mode 10 MB/s (period 4x25 ns, offset 15 bytes)

**Reply by wvalcke** [reply]
Testing with dd using the HD (Fujitsu)
Reading a file with blocksize 64kb -> 9.1MB/s
Writing a file with bloksize 64Kb -> 8.1MB/s

Tests with ZuluSCSI
Writing a file with blocksize 32Kb  -> 4.9MB/s
Writing a file with blocksize 64Kb  -> 4.6MB/s

To me it looks not that bad, but nevertheless, the system boots at last 3 times slower with the ZuluScsi.
Booting the Solaris 9 system will incur a lot of small file reads, maybe this causes the serious drop in performance.

**Reply by aperezbios** [reply, marked as answer]
This is your issue:
 **WARNING: file /HD1.img is not contiguous. This will increase read latency.**

You either need to defragment your image, or copy the image off, reformat your SD card, and copy it back.  Copy the image file before anything else.

**Reply by wvalcke** [reply]
Ok i'll try it. Thanks for the tip.
The way i created the image file was under Linux with the Sd card mounted
truncate -s 15G HD1.img

**Comment by wvalcke** [comment]
OK, some more news, and good news.
I formatted an SD card with Windows, and copied the original files (also via Windows)
I inserted the Sd card and checked the zululog.txt file, the warning that the HD1.img file was non contiguous was gone.
And indeed, the very slow behavior is gone, the system now boots at least as fast as the original HD.
It's probably a bit faster, what i would expect from an SD.
dd if=/dev/zero of=testfile bs=32768 count=2048
Creates a file of 64MB, which is written in 10 seconds. So we have 6.4MB/s the HD is a bit faster, but in general i think the Sd will be faster for reading and seeking. Which is important for normal usage.
Right now i'm also installing Solaris 8 on another img file on the same SD, the install is definitely already a lot faster than before when the image file was 'non contiguous'
I suppose the reason is that the ZuluSCSI firmware detects that the file is non contiguous, it will use the exfat file system for each read/write operation which creates a serious overhead. If the file is contiguous, when the start sector is known for the image on the Sd, the firmware can bypass the exfat filesystem and directly read/write the correct sectors when asked via scsi controller. 
The question i have, i used the instructions as mentioned on the readme. (truncate -s 10G HD1.img)
For some reason this does not create a contiguous file, or maybe due to a bug in the exfat handling of Linux it creates a contiguous file, but flags in the filesystem somewhere it is not. 
I have not yet tested dd if=/dev/zero bs=10G count=1 of=HDx.img, to see if this creates a contiguous file on the SD card.
In any case, it is EXTREMELY IMPORTANT that the ZuluSCSi firmware sees the images as contiguous. 
This cannot be stressed enough in the README to my opinion. 
I will test the ZuluSCSI further on SGI Indy, SGI Indigo2, HP715/100, HP715/60, Sparcstation 1
But to me it looks that the ZuluSCSI performs what i expected, so that makes me happy that i can replace the old SCSi HD's and have an even faster machine due to SD.
Is there maybe a Linux tool which can check is a file is contiguous or not on an exfat partition? filefrag ?
I would like to thank you for the support, this is really appreciated. If you want me to test out more things, let me know.

**Reply by Nable80** [reply]
``truncate -s`` doesn't allocate actual diskspace if your filesystem supports sparse files (FATxx definitely doesn't support it, while exFAT may support, as all modern filesystems do). Try ``fallocate`` or ``dd`` in order to create contiguous files.
You can use ``filefrag -v HD1.img`` to check the result.

**Comment by MeTalOneOEight** [comment]
While you are in experimenting mood, could you try formatting with the official sdcard.org formatter and see if you get better performance?
That's what the sdcard association says and the author of sdfat recommends using it since most OS won't regard the special layout of sd cards.
https://www.sdcard.org/downloads/formatter/

**Reply by wvalcke** [reply]
No problem to test this, but the tools formats the card in FAT32, not in exFAT.
I need exFAT because i need image files bigger than 4GB
Can the tool do an exFAT format ?

**Reply by MeTalOneOEight** [reply]
The manual does not mention it.
Haven't tried if it does it automatically on huge SDs since I don't have any.
But that inspires me for an idea for a new feature:
Segmented image file to work around the 4GB limit if you can't use exFAT.
VMs also use it for their images for better backup, since you only need to copy the changed slice/band instead of the whole image.
